File: src/045. Triangular, pentagonal, and hexagonal/045.java
package test;

import java.util.*;

public class Main
{
    public static void main(String[] args)
    {
        Set<Long> triangle   = new HashSet<>();
        Set<Long> pentagonal = new HashSet<>();
        Set<Long> hexagonal  = new HashSet<>();
        for (long n = 1; n < 100000; n++)
        {
            triangle.add(n * (n + 1) / 2);
            pentagonal.add(n * (3 * n - 1) / 2);
            hexagonal.add(n * (2 * n - 1));
        }
        
        TreeSet<Long> result = new TreeSet<>();
        for (long num: triangle)
        {
            if (pentagonal.contains(num) && hexagonal.contains(num))
            {
                result.add(num);
            }
        }
        
        System.out.println(result.higher((long)40755));
    }
}

File: src/045. Triangular, pentagonal, and hexagonal/045.rb
triangle, pentagonal, hexagonal = [], [], []
1.upto(100000) do |n|
  triangle << n * (n + 1) / 2
  pentagonal << n * (3 * n - 1) / 2
  hexagonal << n * (2 * n - 1)
end

puts (triangle & pentagonal & hexagonal).keep_if {|num| num > 40755}.min
